Tjurina Algebra/Number and finite determinancy for Hypersurfaces

Syz-MS opened the pull request:
Adds 
-Tjurina algebra and number for  Hypersurfaces
-upper determinancy bounds w.r.t. right resp. contact equivalence for IHS
-contact equivalence via Mather-Yau for IHS

(developed at the request of @afkafkafk13)

Syz-MS committed:
Added Code

tjurina algebra, tjurina number, determinancy bounds, contact equivalence  for IHS

Syz-MS committed:
Added Tests

Syz-MS committed:
Fixed some doctests

afkafkafk13 commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+  I = modulus(tjurina_algebra(X, k))
+  if base_ring(I)(1) in I
+    return 0
+  end
```
Maybe you would like to say:

R = tjurina_algebra(X,k)
!is_one(modulus(R)) || return 0

and then use R at later occurrences of the Tjurina algebra

Syz-MS replied:
Removed due to zero dimensional vectorspace case being fixed in #3497

afkafkafk13 commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+    return vector_space_dimension(tjurina_algebra(X, k))
+  catch
+    return -1
+  end
```
You could test for the dimension of the Tjurina algebra and return -1, if it exceeds zero.

fingolfin replied:
This kind of test is problematic because it will intercept *any* error, not just one that indicates something about the dimension.

In general, it would be preferable to not handle this kind of thing via exceptions anyway.

BTW, why does the method just before it not need this "fix" (even though its docstring also says it can return `-1` to indicate infinity)?

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+    tjurina_algebra(X::AffineScheme{<:Field,<:MPolyQuoRing})
+
+Return the global Tjurina algebra of the affine scheme `X`, if `X` is a hypersurface.
+Return an error otherwise.
```
Does it really "return" an error (and what does that mean)? Or does it raise an error / throw an exception?

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+
+
+@doc raw"""
+    tjurina_algebra(X::HypersurfaceGerm)
```
Redundant with the next line
```suggestion
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+@doc raw"""
+    tjurina_number(f::MPolyRingElem)
+
+Return the global Tjurina number of a polynomial `f`. Return -1 if infinite.
```
I really don't like this convention of returning `-1` to indicate infinity. That's gone cause all kinds of bugs down the line.

Alternatives we use elsewhere include:
- return an actual infinite (via `PosInf()`); 
- raise a (dedicated) exception indicating infinite order
- add a separate test function like `has_finite_tjurina_number`

afkafkafk13 replied:
An exception would raise all kinds of complications down the road, I would rather use infinity, as this is the mathematically correct statement.
has_finite_tjurina_number has no added value here, because this is synonymous with is_finitely_determined by the theorem of Mather-Yau and the test for this is available.

Syz-MS replied:
I totally agree with using PosInf() instead of an exception. But I think the problem lies in vector_space_dimension(A::MPolyQuo) throwing an exception instead of returning infinity and therefore it should be fixed at that level. 
Should I open a feature request or fix that issue in this PR?

joschmitt replied:
I wouldn't do it in this pull request.
Tinkering with the output of `vector_space_dimension` would be a breaking change to my understanding and we are not allowed to do these any more since we released 1.0. It would break the code of people who did a similar try-catch work around to catch an infinite dimensional ring if there is no error.
But that's just my two cents.

afkafkafk13 replied:
I would prefer that you open a feature request issue for vectore_space_dimension and use positive infinity here in the meantime, to make sure that everything building on top of it does not break in the future, when the feature is changed.

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+
+
+function _order(f::MPolyRingElem)                                                                                 #TODO:  export???
+  f != 0 || return -1
```
Better:
```suggestion
  !is_zero(f) || return -1
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+function _order(f::MPolyRingElem)                                                                                 #TODO:  export???
+  f != 0 || return -1
+  x = gens(parent(f))
+  n = length(x)
```
No need to allocate `x` if you then don't use it...
```suggestion
  n = ngens(parent(f))
```
or equivalently
```suggestion
  n = nvars(parent(f))
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+  x = gens(parent(f))
+  n = length(x)
+  degrees = [sum([degree(m, j) for j in 1:n]) for m in monomials(f)]
+  return minimum(degrees)
```
... Isn't this just computing the minimal total degrees of all terms? If so something like this should be much more efficient
```suggestion
  return minimum(sum(e) for e in AbstractAlgebra.exponent_vectors(f))
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
````diff
@@ -0,0 +1,770 @@
+true
+```
+"""
+function is_finitelydetermined(f::MPolyLocRingElem{<:Field, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal}, equivalence::Symbol = :contact)
````
Should be like this following our conventions
```suggestion
function is_finitely_determined(f::MPolyLocRingElem{<:Field, <:Any, <:Any, <:Any, <:MPolyComplementOfKPointIdeal}, equivalence::Symbol = :contact)
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+@doc raw"""
+    is_finitelydetermined(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+
+Return if the hypersurface germ 'X' is finitely determined w.r.t. ':right' or ':contact' equivalence. 
```
I would avoid using abbreviations like `w.r.t`, as not everyone knows w.t.m.

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+
+###############################################################################
+@doc raw"""
+    determinacy_bound(f::MPolyLocRingElem)
```
Redundant with next line (the same in a bunch of other docstrings)
```suggestion
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+    k = tjurina_number(f)
+  end
+  characteristic(parent(f)) != 0 || return k + 1
+  characteristic(parent(f)) == 0 || return 2*k - ord_f + 2
```
This seems unnecessarily clever to me. I find this much easier to understand (hopefully I got the order right ... if not I think it kinda underlines my point ;-) )
```suggestion
  return characteristic(parent(f)) == 0 ? k + 1 : 2*k - ord_f + 2
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+  characteristic(R) != 0 || (k = l)
+  ## pos. char.: 2*(l-1) - order_as_series(f) + 2
+  characteristic(R) == 0 || (k = 2*l - order_as_series(f))
+  return k
```
```suggestion
  return characteristic(R) == 0 ? l : 2*l - order_as_series(f)
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+  n != -1 || error("infinite dimensional vectorspaces") 
+  ## Save minimalngens of m^k during calculation
+  ngens_m_k = [length(mA_basis)]  
+  while(length(mA_basis) != n-1) 
```
```suggestion
  while length(mA_basis) != n-1
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+    k += 1
+    gens_mA_k = minimal_generating_set(mA^k)
+    ngens_m_k = vcat(ngens_m_k, length(gens_mA_k))
+    mA_basis = vcat(mA_basis, gens_mA_k)
```
Can't you use `append!` here and above/below?

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+  end
+  ## lift bases
+  mA_basis = [lifted_numerator(mA_basis[j]) for j in 1:length(mA_basis)]  
+  mB_basis = [lifted_numerator(mB_basis[j]) for j in 1:length(mB_basis)]
```
```suggestion
  mA_basis = lifted_numerator.(mA_basis)
  mB_basis = lifted_numerator.(mB_basis)
```

fingolfin commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,770 @@
+    for j in 1:length(mB_basis)
+      img += t[(i*length(mB_basis)+j)]*iota(mB_basis[j])
+    end
+    phi = vcat(phi, img)
```
Consider using `append!` or `push! (as appropriate)

fingolfin commented:
(Just to be clear, overall this PR is fine, the things I pointed out are all minor details and should hopefully to be easy to resolve)

Syz-MS committed:
Apply suggestions from code review

Co-authored-by: Max Horn <max@quendi.de>

Syz-MS committed:
Merge branch 'oscar-system:master' into ms/tjurina

Syz-MS committed:
Added more suggestions from code review

is_finitelydetermined -> is_finitely_determined
w.r.t. -> with respect to
vcat -> append! resp. push!
and minor changes

Syz-MS commented:
@fingolfin Thank you for your suggestions. They have been quite helpful.

Syz-MS committed:
using PosInf() instead of -1

tjurina number now returns infinity instead of -1
added more test
using smaller k-th tjurina algebra for Mather-Yau in pos. char. to reduce computing effort

afkafkafk13 commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,751 @@
+    determinacy_bound(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+
+Compute some determinacy bound of the hypersurface germ 'X' with respect to ':right' or ':contact' equivalence.
+Return infinity if not finitely determined. 
```
Please state explicitly somewhere that PosInf() does not prove the absence of finite determinacy in prime characteristic. In positive characteristic infinity is also returned, if no determinacy bound can be obtained by this approach.

Syz-MS replied:
In my understanding an upper determinancy bound (strictly less than infinity) can always be computed this way in prime characteristic acording to the paper ''Invariants of Hypersurface Singularities in Positive Characteristic'' by Boubakri, Greuel, Markwig (arXiv:1005.4503). 
The upper bounds provided by Theorem 2.1 respectivly Corollary 2.4 are computated here and as a consequence of Theorem 2.5 these bounds are always finite in the finite determined case. See also Theorem 2.8. Therefore PosInf() is only returned in the non finite determined case.

afkafkafk13 replied:
Yes indeed. I take my previous comment back (I had some other invariant in mind).

afkafkafk13 commented on experimental/Schemes/Tjurina.jl:
```diff
@@ -0,0 +1,751 @@
+    sharper_determinacy_bound(X::HypersurfaceGerm, equivalence::Symbol = :contact)
+
+Compute some determinacy bound of the hypersurface germ 'X' with respect to ':right' or ':contact' equivalence.
+Return infinity if not finitely determined. 
```
dito

afkafkafk13 reviewed (commented):
add the two small remarks to the documentation, then this is good to go

afkafkafk13 committed:
Merge branch 'master' into ms/tjurina

Syz-MS committed:
Merge branch 'oscar-system:master' into ms/tjurina

afkafkafk13 commented:
Concerning the tests: The structure of including a file in experimental has changed. You need to add your file in `experimental/Schemes/Schemes.jl`

Syz-MS committed:
Adapting to #3409
